### PR TITLE
Add ability to change billing account for a billing project

### PR DESCRIPTION
**Pull request by petesantos**
Added calls to google API's to fetch and update the billing account for a project
Added a select box to choose a billing account on the billing projects page
Fetch the available billing accounts at page load time and pass the list as props to the components using this information

### CODE REVIEW

File: src/libs/ajax.js [hunk 1]
```diff
@@ -765,6 +765,20 @@ const GoogleBilling = signal => ({
     const response = await fetchGoogleBilling(`${billingAccountName}/projects`, _.merge(authOpts(), { signal }))
     const json = await response.json()
     return _.map('projectId', json.projectBillingInfo)
+  },
+  getBillingInfo: async project => {
+    const response = await fetchGoogleBilling(`projects/${project}/billingInfo`, _.merge(authOpts(), { signal }))
+    return await response.json()
```
Comment (panentheos) [inline src/libs/ajax.js]: nit: `await` is unnecessary here due to promise collapsing, you can just `return response.json()`

File: src/libs/ajax.js [hunk 2]
```diff
@@ -765,6 +765,20 @@ const GoogleBilling = signal => ({
...
+  },
+  changeBillingAccount: async ({ projectId, newAccountName }) => {
+    const name = `projects/${projectId}/billingInfo`
+    const billingAccountName = `${newAccountName}`
+    const response = await fetchGoogleBilling(`projects/${projectId}/billingInfo`, _.merge(authOpts(), {
+      signal,
+      method: 'PUT',
+      body: JSON.stringify({ billingEnabled: true, billingAccountName, name, projectId })
```
Comment (panentheos) [inline src/libs/ajax.js]: There's a `jsonBody` helper that will add the stringified body plus the right content-type header. (Will need to switch to `_.mergeAll` in that case)

File: src/pages/billing/List.js [hunk 3]
```diff
@@ -200,26 +186,35 @@ export const BillingList = _.flow(
...
+       _.map(gs => ({ ...gs[0], role: _.map('role', gs) })),
+       _.sortBy('projectName')
+     )(rawBillingProjects)
+     this.setState({ billingProjects })
+   })
+
+  loadAccounts = withErrorReporting('Error loading billing accounts',
+    (async () => {
+      const { ajax: { Billing } } = this.props
+      await Auth.ensureBillingScope()
```
Comment (panentheos) [inline src/pages/billing/List.js]: Moving this to the top level means that the page will ask for billing scopes immediately on navigation, right? We should check that that's acceptable.

File: src/pages/billing/List.js [hunk 4]
```diff
@@ -200,26 +186,35 @@ export const BillingList = _.flow(
...
+     const billingProjects = _.flow(
+       _.groupBy('projectName'),
+       _.map(gs => ({ ...gs[0], role: _.map('role', gs) })),
+       _.sortBy('projectName')
+     )(rawBillingProjects)
+     this.setState({ billingProjects })
+   })
+
+  loadAccounts = withErrorReporting('Error loading billing accounts',
+    (async () => {
```
Comment (panentheos) [inline src/pages/billing/List.js]: nit: unnecessary parens

File: src/pages/billing/Project.js [hunk 5]
```diff
@@ -22,44 +21,72 @@ export default ajaxCaller(class ProjectDetail extends Component {
...
-    this.setState({ addingUser: false, deletingUser: false, updating: false, editingUser: false })
-    const rawProjectUsers = await Billing.project(project.projectName).listUsers()
-    const projectUsers = _.flow(
-      _.groupBy('email'),
-      _.map(gs => ({ ..._.omit('role', gs[0]), roles: _.map('role', gs) })),
-      _.sortBy('email')
-    )(rawProjectUsers)
-    this.setState({ projectUsers })
+  )(async newAccountName => {
+    const { ajax: {  GoogleBilling }, project: { projectName }  } = this.props
```
Comment (panentheos) [inline src/pages/billing/Project.js]: nit: extra spaces

File: src/pages/billing/Project.js [hunk 6]
```diff
@@ -22,44 +21,72 @@ export default ajaxCaller(class ProjectDetail extends Component {
...
             () => icon('error-standard', { style: { color: colors.danger() } })
-          )
+          ),
+          span({ style: { fontWeight: 500, fontSize: 14, margin: '0 0.75rem 0 auto' } }, 'Billing Account:'),
+          h(Select, {
+            value: billingAccountName,
+            isClearable: false,
+            styles: { container: old => ({ ...old, width: 320 }) },
+            options: accounts,
+            onChange: async ({ value: newAccountName }) => this.updateBillingAccount(newAccountName)
```
Comment (panentheos) [inline src/pages/billing/Project.js]: nit: `async` is unnecessary since we're not using `await`. Also, no need to rename `value`.

File: src/libs/ajax.js (hunk 1 above)
Comment (petesantos) [reply src/libs/ajax.js]: removed - thanks

File: src/pages/billing/List.js (hunk 4 above)
Comment (petesantos) [reply src/pages/billing/List.js]: fixed

Comment (petesantos) [reply src/pages/billing/List.js]: fixed

Comment (petesantos) [reply src/pages/billing/List.js]: fixed

File: src/pages/billing/Project.js (hunk 5 above)
Comment (petesantos) [reply src/pages/billing/Project.js]: fixed

File: src/pages/billing/Project.js (hunk 6 above)
Comment (petesantos) [reply src/pages/billing/Project.js]: fixed

File: src/libs/ajax.js (hunk 2 above)
Comment (petesantos) [reply src/libs/ajax.js]: Updated to use the helper - thanks!

File: src/pages/billing/List.js (hunk 3 above)
Comment (petesantos) [reply src/pages/billing/List.js]: I spoke with Brad, looking into this further to see what can be done to allow users to continue to modify the data they area authorized to change when they are not authorized to change the billing account.

Comment (petesantos) [reply src/pages/billing/List.js]: I spoke with Brad, looking into this further to see what can be done to allow users to continue to modify the data they area authorized to change when they are not authorized to change the billing account.

Comment (petesantos) [reply src/pages/billing/List.js]: I spoke with Brad, looking into this further to see what can be done to allow users to continue to modify the data they area authorized to change when they are not authorized to change the billing account.

File: config-overrides.js [hunk 7]
```diff
@@ -13,6 +13,13 @@ module.exports = {
 
     config = _.merge(config, manualOverrides)
 
+    config.module.rules[2].oneOf.unshift(
```
Comment (katevoss) [inline config-overrides.js]: Why did you add this? It looks like you need to update from dev on your branch.

Comment (petesantos) [reply config-overrides.js]: I think I corrected the issue with the extraneous files that were added in a rebase I had done.
I am now seeing 3 files changed (as it should be). If you still see this please flag it again - thanks!

File: src/pages/billing/List.js (hunk 3 above)
Comment (petesantos) [reply src/pages/billing/List.js]: I updated the logic so the authorization scope isn't requested until necessary - either adding or changing the billing account

File: src/libs/ajax.js [hunk 8]
```diff
@@ -811,6 +811,20 @@ const GoogleBilling = signal => ({
...
     const json = await response.json()
     return _.map('projectId', json.projectBillingInfo)
+  },
+  getBillingInfo: async project => {
+    const response = await fetchGoogleBilling(`projects/${project}/billingInfo`, _.merge(authOpts(), { signal }))
+    return response.json()
+  },
+  changeBillingAccount: async ({ projectId, newAccountName }) => {
+    const name = `projects/${projectId}/billingInfo`
+    const billingAccountName = `${newAccountName}`
```
Comment (breilly2) [inline src/libs/ajax.js]: nit: unnecessary `const`

File: src/pages/billing/List.js [hunk 9]
```diff
@@ -196,30 +181,50 @@ export const BillingList = _.flow(
...
+      billingAccounts: null,
+      billingScope: false,
       ...StateHistory.get()
     }
   }
 
-  async componentDidMount() {
-    this.loadProjects()
-  }
+  componentDidMount = Utils.withBusyState(v => this.setState({ isBusy: v }),
```
Comment (breilly2) [inline src/pages/billing/List.js]: Super minor nit/question that I normally wouldn't mention in a PR but would quickly ask if we were mobbing...

How about:
```
Utils.withBusyState(isBusy => this.setState({ isBusy }))
```
to avoid introducing a separate identifier `v`?

File: src/pages/billing/List.js [hunk 10]
```diff
@@ -196,30 +181,50 @@ export const BillingList = _.flow(
...
-    const { ajax: { Billing } } = this.props
-    const rawBillingProjects = await Billing.listProjects()
-    const billingProjects = _.flow(
-      _.groupBy('projectName'),
-      _.map(gs => ({ ...gs[0], role: _.map('role', gs) })),
-      _.sortBy('projectName')
-    )(rawBillingProjects)
-    this.setState({ billingProjects })
+    await Auth.ensureBillingScope()
+    await this.loadAccounts()
```
Comment (breilly2) [inline src/pages/billing/List.js:210]: TOL: I think a function named `authorizeBillingScope` shouldn't also be loading the billing accounts.

File: src/pages/billing/List.js [hunk 11]
```diff
@@ -196,30 +181,50 @@ export const BillingList = _.flow(
...
   })
 
+  loadAccounts = withErrorReporting('Error loading billing accounts',
+    async () => {
+      const { ajax: { Billing } } = this.props
+      if (await Auth.hasBillingScope()) {
```
Comment (breilly2) [inline src/pages/billing/List.js]: No need to `await` here. `Auth.hasBillingScope()` is not `async`.

File: src/pages/billing/List.js [hunk 12]
```diff
@@ -239,9 +244,13 @@ export const BillingList = _.flow(
         div({ style: { width: 330, boxShadow: '0 2px 5px 0 rgba(0,0,0,0.25)' } }, [
           div({ style: Style.navList.heading }, [
             'Billing Projects',
-            h(Clickable,
-              { onClick: () => { this.setState({ creatingBillingProject: true }) } },
-              [icon('plus-circle', { size: 21, style: { color: colors.accent() } })]
+            h(Clickable, {
+              onClick: () => {
+                this.authorizeBillingScope()
```
Comment (breilly2) [inline src/pages/billing/List.js]: Suggestion to prevent having a double busy spinner (which is especially awkward if you close the pop-up without granting billing scope):
```
await this.authorizeBillingScope()
if (Auth.hasBillingScope()) {
  this.setState({ creatingBillingProject: true })
} else {
  // nothing to do here, as long as authorizeBillingScope() caught the error and notified the user
}
```

File: src/pages/billing/List.js [hunk 13]
```diff
@@ -196,30 +181,50 @@ export const BillingList = _.flow(
...
+    await Auth.ensureBillingScope()
+    await this.loadAccounts()
   })
 
+  loadAccounts = withErrorReporting('Error loading billing accounts',
+    async () => {
+      const { ajax: { Billing } } = this.props
+      if (await Auth.hasBillingScope()) {
+        const billingAccounts = await Billing.listAccounts()
+        this.setState({ billingScope: true, billingAccounts })
```
Comment (breilly2) [inline src/pages/billing/List.js]: I'm suspicious of the need for `billingScope` in state. We should be able to call `Auth.hasBillingScope()` as needed to get this.

**Review by breilly2 — COMMENTED** [review]
Some initial thoughts. I might have more but I don't want to hold these comments up, and subsequent changes might affect what else I have to say anyway.

File: src/libs/ajax.js (hunk 8 above)
Comment (petesantos) [reply src/libs/ajax.js]: fixed - thanks

File: src/pages/billing/List.js (hunk 9 above)
Comment (petesantos) [reply src/pages/billing/List.js]: I was keeping with the existing style of the code - but I agree that looks much cleaner - fixed

File: src/pages/billing/List.js (hunk 11 above)
Comment (petesantos) [reply src/pages/billing/List.js]: Thanks - I was thinking this was a network request - fixed.

File: src/pages/billing/List.js (hunk 13 above)
Comment (petesantos) [reply src/pages/billing/List.js]: This is a result from my incorrect understanding of hasBillingScope being async - removed the state.

File: src/pages/billing/List.js (hunk 12 above)
Comment (petesantos) [reply src/pages/billing/List.js]: Used this logic with a slight change to prevent any async calls if the billing scope is already set - thanks!

File: src/pages/billing/List.js (hunk 10 above)
Comment (petesantos) [reply src/pages/billing/List.js:210]: I renamed this function, thanks.

File: src/pages/billing/Project.js [hunk 14]
```diff
@@ -22,44 +22,82 @@ export default ajaxCaller(class ProjectDetail extends Component {
...
+    ])
+  )
 
   render() {
-    const { project: { projectName, creationStatus }, ajax: { Billing } } = this.props
-    const { projectUsers, loading, updating, filter, addingUser, deletingUser, editingUser } = this.state
+    const { project: { projectName, creationStatus }, ajax: { Billing }, billingAccounts, authorizeAndLoadAccounts, loadingAccountAuth } = this.props
+    const { projectUsers, loading, updating, filter, addingUser, deletingUser, editingUser, billingAccountName } = this.state
     const adminCanEdit = _.filter(({ roles }) => _.includes('Owner', roles), projectUsers).length > 1
+    const billingNode = Utils.cond(
```
Comment (katevoss) [inline src/pages/billing/Project.js]: Nit: `billingNode` is a little general, maybe rename it to something more specific like `selectBillingAccount`?

File: src/pages/billing/Project.js [hunk 15]
```diff
@@ -22,44 +22,82 @@ export default ajaxCaller(class ProjectDetail extends Component {
...
+        div({ style: { color: colors.dark(), fontSize: 16, fontWeight: 600, display: 'flex', alignItems: 'center' } }, [
           projectName,
           span({ style: { fontWeight: 500, fontSize: 14, margin: '0 1.5rem 0 3rem' } }, creationStatus),
           Utils.cond(
             [creationStatus === 'Ready', () => icon('check', { style: { color: colors.success() } })],
             [creationStatus === 'Creating', () => spinner({ size: 16 })],
             () => icon('error-standard', { style: { color: colors.danger() } })
-          )
+          ),
+          billingNode && span({ style: { flexShrink: 0, fontWeight: 500, fontSize: 14, margin: '0 0.75rem 0 auto' } }, 'Billing Account For This Project:'),
```
Comment (katevoss) [inline src/pages/billing/Project.js]: Why is `billingNode` used as a conditional here?

Comment (petesantos) [reply src/pages/billing/Project.js]: Originally, I only wanted the label/span to display once the type of node was determined (which was async). I now use hasBillingScope for this which is synchronous - thanks!

File: src/pages/billing/List.js [hunk 16]
```diff
@@ -192,30 +177,49 @@ export const BillingList = _.flow(
...
+      billingAccounts: null,
       ...StateHistory.get()
     }
   }
 
-  async componentDidMount() {
-    this.loadProjects()
-  }
+  componentDidMount = Utils.withBusyState(isBusy => this.setState({ isBusy }),
+    () => Promise.all([
```
Comment (breilly2) [inline src/pages/billing/List.js]: Does this need the `Promise.all`? From the last time I did something like this, I remember this stuff being fiddly. However I'm a little suspicious about the bundling of these into a single promise that is not `await`ed.

File: src/pages/billing/List.js [hunk 17]
```diff
@@ -251,14 +263,22 @@ export const BillingList = _.flow(
...
             this.loadProjects()
           }
         }),
-        !!selectedName && billingProjects && h(ProjectDetail, { key: selectedName, project: _.find({ projectName: selectedName }, billingProjects) }),
-        isBusy && spinnerOverlay
+        !!selectedName && billingProjects && h(ProjectDetail, {
+          key: selectedName,
+          project: _.find({ projectName: selectedName }, billingProjects),
+          billingAccounts,
+          hasBillingScope: Auth.hasBillingScope(),
```
Comment (breilly2) [inline src/pages/billing/List.js]: This prop seems unnecessary. `ProjectDetail` should be able to use `Auth.hasBillingScope()` directly.

File: src/pages/billing/Project.js [hunk 18]
```diff
@@ -22,44 +22,73 @@ export default ajaxCaller(class ProjectDetail extends Component {
...
             () => icon('error-standard', { style: { color: colors.danger() } })
-          )
+          ),
+          span({ style: { flexShrink: 0, fontWeight: 500, fontSize: 14, margin: '0 0.75rem 0 auto' } }, 'Billing Account For This Project:'),
+          hasBillingScope ? h(Select, {
+            value: billingAccountName,
+            isClearable: false,
+            styles: { container: old => ({ ...old, width: 320 }) },
+            options: _.map(({ displayName, accountName }) => ({ label: displayName, value: accountName }), billingAccounts),
+            onChange: ({ value: newAccountName }) => billingAccountName !== newAccountName && this.updateBillingAccount(newAccountName)
```
Comment (breilly2) [inline src/pages/billing/Project.js]: This feels like a heavy operation to have `onChange`. I don't think we have precedent in Terra for making changes like this without a confirm/cancel. Based on the current layout, I don't have great suggestions about how to do this. Thoughts?

**Review by breilly2 — CHANGES_REQUESTED** [review]
A couple of nits, but also a UX thing that I think we should change, or at least be very conscious of the deviation from what we've historically done in Terra.

File: src/pages/billing/List.js (hunk 17 above)
Comment (petesantos) [reply src/pages/billing/List.js]: The reason I am using this as a prop here is because the detail page does not know when the scope has changed. The scope changes are handled in List.js (at the page level). I could push this down to the Project Detail component, but the project detail would then need to track these changes in componentDidUpdate which can get messy as it would need to track the previous state of the billing scope and possibly some other additional state.

File: src/pages/billing/Project.js (hunk 18 above)
Comment (petesantos) [reply src/pages/billing/Project.js]: I can move this to a modal, but it would impact UX as it would be a lot of clicks to make a change. I feel like we should mob on this.

File: src/pages/billing/List.js (hunk 16 above)
Comment (petesantos) [reply src/pages/billing/List.js]: What were the issues? Promise.all is used throughout the code base. When I don't wait on both requests, the billing account will just "pop in" to the select, which yields a slightly jarring UX. On the await, wrapping this function in withBusyState adds an async await to the function. I can add an additional await for better clarity if needed

File: src/pages/billing/List.js [hunk 19]
```diff
@@ -192,30 +177,54 @@ export const BillingList = _.flow(
...
       creatingBillingProject: false,
+      billingAccounts: null,
       ...StateHistory.get()
     }
   }
 
-  async componentDidMount() {
-    this.loadProjects()
+  componentDidMount() {
+    Promise.all([
```
Comment (breilly2) [inline src/pages/billing/List.js]: Chatted in person. Can remove this `Promise.all` since nothing is `await`ing it since each of the contained promises handles their own busy/error.

(Same for `ProjectDetail.componentDidMount`)

**Review by breilly2 — APPROVED** [review]
👍 after a minor change discussed in person
